book/invalidation.md

```diff
@@ -0,0 +1,379 @@
+other types of layout objects and `append`s them to its `children`
+array. Here, the easy fix is to reset the `children` array at the top
+of `layout`:
+
+``` {.python}
+class BlockLayout:
+    def layout(self, zoom):
+        self.children = []
+        # ...
+```
+
+Now the `BlockLayout`'s `layout` function is idempotent again, because
+each call will recreate the `children` array the same way each time.
+
+I examined all of the other `layout` methods to check whether they are
+idempotent. I found a few worrying causes:
+
+- In `new_line`, `BlockLayout` will append to its `children` array;
+- In `text` and `input`, `BlockLayout` will append to the `children`
+  array of some `LineLayout` child
+- In `text` and `input`, `BlockLayout` will call `get_font`, as will
+  the `TextLayout` and `InputLayout` methods
+- Basically every layout method calls `display_px`
+
+Luckily, none of these break idempotency. The `new_line` method is
+only called through `layout`, which resets the `children` array.
+Similarly, `text` and `input` are also only called through `layout`,
+which means all of the `LineLayout` children are reset. And `get_font`
+acts as a cache, so multiple calls return the same font object, while
+`display_px` just does some math so always returns the same result
+given the same inputs. This means the browser should now work
+correctly again.
+
+::: {.further}
+
+:::
+
+Dependencies
+============
+
+So far, we've only looked at one place where layout objects are
+created. Let's look at another: `BlockLayout`s created by
+`BlockLayout`s in their `layout` method. Here's the relevant code;
+it only runs in block layout mode:
+
+``` {.python}
+class BlockLayout:
+    def layout(self, zoom):
+        self.children = []
+        # ...
+        if layout_mode(self.node) == "block":
+            previous = None
+            for child in self.node.children:
+                next = BlockLayout(child, self, previous)
+                self.children.append(next)
+                previous = next
+        # ...
+```
+
+Here, the arguments to `BlockLayout` are a little more complicated
+than the ones passed in `DocumentLayout`. The second argument, `self`,
+is never assigned to, but `child` and `previous` both contain elements
+read from `node.children`, and that `children` array can change as a
+result of `innerHTML`. Moreover, in order to even run this code, the
+node's `layout_mode` has to be `"block"`, and `layout_mode` itself
+also reads the node's `children`.[^and-tags]
+
+[^and-tags]: It also looks at the node's `tag` and the node's
+    childrens' `tag`s, but luckily in our browser a node's `tag` can't
+    change, so we don't need to worry about this. If our browser
+    supported changing a node's `tag`, that's something we would need
+    to worry about, but to my knowledge no real browser does that,
+    even in APIs like `outerHTML`.
+
+So to avoid redundant `BlockLayout` creation, we're going to have to
```

Comment:
I think you should tie this back to idempotence more clearly:
- Idempotence means you can store off the result of the first function call in a cache 
- The functions should be idempotent (and if not you should go edit them to be so), but that doesn't mean the output doesn't change if the input does
- When the input changes, you have to pair idempotence of the function with a dirty flag indicating the input might have changed.
- You might want more than one dirty flags to reduce the likelihood of an invalidation

These together form the basis for a layout invalidation strategy.

These points together are a lot to understand at once, so at this point I think the first 3 bullets and how they are related are the important things to explain.

Reply:
Maybe this justifies pseudocode before getting to the main action?

function idempotent(arg1, arg2) {
  //
}

function run_idempotent(arg1, arg2) {
  if (not saved)
    saved = idempotent(arg1, arg2)
  return saved
}

if (arg1 changes or arg2 changes)
   saved = undefined

Then if this is super clear, we can point out that the really difficult thing is that not everything is an explicit argument to a function, and sometimes variables are stored off here and there. It's easy to forget they are in fact implicit arguments, and also easy to forget their dependencies in turn.

you say all this below, just wondering out loud if the pseudocode is helpful to explain that the core concept is actually very simple, it's the implementation in a complex code base without making really slow or unreadable, that is  hard.

Maybe not, just an idea.

Reply:
I added a discussion of invalidation vs dirty flags. I think it's a really helpful point and improves the flow.

I'm going to add a discussion of implicit arguments once we get to actual layout invalidation, because that's where we first see them. I've went back and rewrote the definition of idempotence to be "same output if called with the same inputs *and dependencies*", and then go back to that in sec 2.

I'm hesitant to do pseudo-code because that introduces a whole extra layer of mapping the pseudo-code to the actual code. Let me write more of the chapter and see if it would help or hurt. I did consider some kind of abstract `ProtectedField` class, where you explicitly list the inputs & outputs and it does the dirty flag stuff. That way the "pseudo-code" becomes actual code. But the layout code ended up looking kind of ugly and I either needed to define a class per field or use lambdas and other more Python-specific features, neither of which I was thrilled with. But happy to discuss!